abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,919 @@
+  - `i32 (size_t) buffer_size`
+  - `i32 (uint32_t) flags`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set content of the buffer `buffer_type` to the bytes (`buffer_data`, `buffer_size`), replacing
+`size` bytes, starting at `offset` in the existing buffer.
+
+
+### `proxy_get_map`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (const char**) return_map_data`
+  - `i32 (size_t*) return_map_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get all key-value pairs from a given map (`map_type`).
+
+
+### `proxy_set_map`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (const char*) map_data`
+  - `i32 (size_t) map_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set all key-value pairs in a given map (`map_type`).
+
+
+### `proxy_get_map_value`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const char**) return_value_data`
+  - `i32 (size_t*) return_value_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get content of key (`key_data`, `key_size`) from a given map (`map_type`).
+
+
+### `proxy_set_map_value`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const char*) value_data`
+  - `i32 (size_t) value_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set or replace the content of key (`key_data`, `key_size`) to the value (`value_data`, `value_size`)
+in a given map (`map_type`).
+
+### `proxy_add_map_value`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const char*) value_data`
+  - `i32 (size_t) value_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Add key (`key_data`, `key_size`) with the value (`value_data`, `value_size`) to a given map
+(`map_type`).
+
+
+### `proxy_remove_map_value`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Remove key (`key_data`, `key_size`) from a given map (`map_type`).
+
+
+### `proxy_get_property`
+
+* params:
+  - `i32 (const char*) property_path_data`
+  - `i32 (size_t) property_path_size`
+  - `i32 (const char**) return_property_value_data`
+  - `i32 (size_t*) return_property_value_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get property (`property_path_data`, `property_path_size`).
+
+
+### `proxy_set_property`
+
+* params:
+  - `i32 (const char*) property_path_data`
+  - `i32 (size_t) property_path_size`
+  - `i32 (const char*) property_value_data`
+  - `i32 (size_t) property_value_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set property (`property_path_data`, `property_path_size`) to given value (`property_value_data`,
+`property_value_size`).
+
+
+## TCP/UDP/QUIC stream (L4) extensions
+
+### `proxy_resume_downstream`
```

Comment:
Let's combine these into one proxy_resume(ResumeType) like we have done with the other calls.

In general I think we need enum ResumeType {
  Downstream,
  Upstream,
  Request?
  Response?
}

In the future it is likely that we will want to be able to do the sort of merging of multiple upstream sources into a single downstream flow.   In this case we could extend the enum to include

Stream0 = 100,
Stream99 = 199,

To represent other streams which might be opened by the plugin.

We could also use this for backpressure on say gRPC streams.